enzyme/Enzyme/MLIR/Passes/EnzymeToMemRef.cpp

```diff
@@ -150,6 +150,122 @@ struct LoweredCache {
+                                           OpBuilder &b) const {
+    MLIRContext *context = b.getContext();
+    std::string funcName = "__enzyme_set";
+    llvm::raw_string_ostream funcStream{funcName};
+    funcStream << elementType;
+    if (moduleOp.lookupSymbol<func::FuncOp>(funcName)) {
+      return SymbolRefAttr::get(context, funcName);
+    }
+
+    OpBuilder::InsertionGuard insertionGuard(b);
+    b.setInsertionPointToStart(moduleOp.getBody());
+
+    auto setFnType = FunctionType::get(
+        context, /*inputs=*/
+        {elements.getType(), size.getType(), capacity.getType(), elementType},
+        /*outputs=*/{});
+    auto setFn = b.create<func::FuncOp>(loc, funcName, setFnType);
+    setFn.setPrivate();
+    Block *entryBlock = setFn.addEntryBlock();
+    b.setInsertionPointToStart(entryBlock);
+    BlockArgument elementsField = setFn.getArgument(0);
+    BlockArgument sizeField = setFn.getArgument(1);
+    BlockArgument capacityField = setFn.getArgument(2);
+    BlockArgument value = setFn.getArgument(3);
+
+    Value sizeVal = b.create<memref::LoadOp>(loc, sizeField);
+    Value capacityVal = b.create<memref::LoadOp>(loc, capacityField);
+
+    Value predicate = b.create<arith::CmpIOp>(loc, arith::CmpIPredicate::eq,
+                                              sizeVal, capacityVal);
+    b.create<scf::IfOp>(
+        loc, predicate, [&](OpBuilder &thenBuilder, Location loc) {
+          Value two = thenBuilder.create<arith::ConstantIndexOp>(loc, 2);
+          Value newCapacity =
+              thenBuilder.create<arith::MulIOp>(loc, capacityVal, two);
+          Value oldElements =
+              thenBuilder.create<memref::LoadOp>(loc, elementsField);
+          Value newElements = thenBuilder.create<memref::AllocOp>(
+              loc, oldElements.getType().cast<MemRefType>(), newCapacity);
+          thenBuilder.create<memref::CopyOp>(loc, oldElements, newElements);
+          thenBuilder.create<memref::DeallocOp>(loc, oldElements);
+          thenBuilder.create<memref::StoreOp>(loc, newElements, elementsField);
+          thenBuilder.create<memref::StoreOp>(loc, newCapacity, capacityField);
+          thenBuilder.create<scf::YieldOp>(loc);
+        });
+
+    Value elementsVal = b.create<memref::LoadOp>(loc, elementsField);
+    b.create<memref::StoreOp>(loc, value, elementsVal,
+                              /*indices=*/sizeVal);
+
+    Value one = b.create<arith::ConstantIndexOp>(loc, 1);
+    Value newSize = b.create<arith::AddIOp>(loc, sizeVal, one);
+
+    b.create<memref::StoreOp>(loc, newSize, sizeField);
+    b.create<func::ReturnOp>(loc);
+
+    return SymbolRefAttr::get(context, funcName);
+  }
+
+  void emitSet(Location loc, Value value, OpBuilder &b,
+               FlatSymbolRefAttr setFn) const {
+    b.create<func::CallOp>(
+        loc, setFn, /*results=*/TypeRange{},
+        /*operands=*/ValueRange{elements, size, capacity, value});
+  }
```

Comment:
@umatin is there a case where we would need to `set` a cache? Would this be different than a push?

Reply:
I think currently this should not be happening, but I think having a set op write to the top element instead of doing a push would be more intuitive and might be useful later on. I would suggest we do not lower set on cache at the moment until we need it.